firmware/phy_task.c

```diff
@@ -0,0 +1,718 @@
+    enable_irq(MCU_FAST_IRQ_INIT_START);
+}
+
+/*-----------------------------------------------------------*/
+static void fsw_post_switch_handler(__UNUSED__ void *currentStateData,
+                                    __UNUSED__ struct event *event,
+                                    __UNUSED__ void *newStateData)
+{
+    // Indicate to PLL that switch occurred; disable previous vco
+    pll_switch_vco(&wddr.pll, false);
+    pll_disable_vco(&wddr.pll);
+
+    // Turn off pll interrupt
+    pll_set_lock_interrupt_state(&wddr.pll, false);
+
+    // Indicate Post Work Done
+    fsw_ctrl_set_post_work_done(false, 0x1);
+
+    // Put back in default state
+    fsw_ctrl_set_post_work_done(false, 0x0);
+    fsw_ctrl_set_prep_done_reg_if(false);
+    fsw_ctrl_set_msr_toggle_en_reg_if(0x1);
+    fsw_ctrl_set_vco_toggle_en_reg_if(0x1);
+}
+
+/*-----------------------------------------------------------*/
+static void fsw_pending_entry_handler(__UNUSED__ void *stateData,
+                                      __UNUSED__ struct event *event)
+{
+    uint8_t init_start, init_complete;
+
+    // Wait for INIT_START to go low
+    do
+    {
+        init_start = dfi_get_init_start_status_reg_if();
+    } while (init_start != 0x0);
+
+    // Override Init Complete and force low to Complete MRW
+    dfi_set_init_complete_ovr_reg_if(true, 0x1);
+
+    // PHY should have completed frequency switch and indicated init_complete
+    do
+    {
+        init_complete = dfi_get_init_complete_status_reg_if();
+    } while (init_complete != 0x1);
+
+    // Set Post Work Done Override
+    fsw_ctrl_set_post_work_done(true, 0x0);
+
+    // Send MRW from DFI Buffer (This was filled during prep)
+    dfi_buffer_send_packets(&wddr.dfi.dfi_buffer, false);
+
+    // Must disable buffer when done
+    dfi_buffer_disable(&wddr.dfi.dfi_buffer);
+
+    // Release Init Complete Override
+    dfi_set_init_complete_ovr_reg_if(false, 0x1);
+
+    // Allow PLL lock interrupt
+    pll_set_lock_interrupt_state(&wddr.pll, true);
+}
+
+/*-----------------------------------------------------------*/
+static void dfi_phymstr_pending_entry_handler(__UNUSED__ void *stateData,
+                                              struct event *event)
+{
+    // Get master signal states from event data
+    dfi_phymstr_req_t *req = (dfi_phymstr_req_t *) event->data;
+
+    // Request PHY MASTER control
+    dfi_phymstr_req(&wddr.dfi, req);
+
+    // Enable PHY MASTER ACK IRQ
+    enable_irq(MCU_FAST_IRQ_PHYMSTR_ACK);
+}
+
+/*-----------------------------------------------------------*/
+static void dfi_phymstr_enter_handler(void *stateData, struct event *event)
+{
+#if CONFIG_CAL_PERIODIC
+    vComplete(&fw_manager.enteredPhyMaster);
+#endif /* CONFIG_CAL_PERIODIC */
+}
+
+/*-----------------------------------------------------------*/
+static void dfi_phymstr_exit_handler(void *stateData, struct event *event)
+{
+    // TODO: Should DRAM be put back into correct state or done elsewhere?
```

Comment:
I believe that once we release phy master control, the MC will expect DRAM to be in the same state it was when we entered master.  I would expect we would place DRAM back into the correct state after changes prior to calling exit function.

Reply:
Yes, it is supposed to do that hints the TODO. The TODO is just asking should it be done here or should it be done somewhere else.

The TODO is leftover from the original DFI code that I wrote last year. Back then, I didn't want DRAM code mixed into the DFI FSM so I left that TODO there to revisit it later. With the changes you made, I think it's easier to integrate DRAM calls than it was then. However, I'm still not sure I want the code integrated here or not.

The caller that request PHY MASTER control will be wanting to use the DRAM. I'm afraid of sticking code here to put it back into the original state because I'm worried we might have to do different sequences based on what was done to the DRAM during PHY MASTER control. Given that, I have always fallen on the side that the caller could put the DRAM back into the state more safely (i.e. it could know exact sequence / timing it needs to perform) given that it knows what it has done to the DRAM during PHY MASTER control. It would need to put it back to the original state, then request PHY MASTER exit.